aptos-move/framework/aptos-framework/sources/Voting.move

```diff
@@ -0,0 +1,317 @@
+        voting_forum.next_proposal_id = voting_forum.next_proposal_id + 1;
+
+        Table::add(&mut voting_forum.proposals, proposal_id, Proposal {
+            execution_content: Option::some<ProposalType>(execution_content),
+            execution_hash,
+            min_vote_threshold,
+            expiration_secs,
+            early_resolution_vote_threshold,
+            yes_votes: 0,
+            no_votes: 0,
+            resolved: false,
+        });
+        proposal_id
+    }
+
+    /// Vote on the given proposal.
+    ///
+    /// @param _proof Required so only the governance module that defines ProposalType can initiate voting.
+    ///               This guarantees that voting eligibility and voting power are controlled by the right governance.
+    /// @param voting_forum_address The address of the forum where the proposals are stored.
+    /// @param proposal_id The proposal id.
+    /// @param num_votes Number of votes. Voting power should be calculated by governance.
+    /// @param should_pass Whether the votes are for yes or no.
+    public fun vote<ProposalType: store>(
+        _proof: &ProposalType,
+        voting_forum_address: address,
+        proposal_id: u64,
+        num_votes: u128,
+        should_pass: bool,
+    ) acquires VotingForum {
+        let voting_forum = borrow_global_mut<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = Table::borrow_mut(&mut voting_forum.proposals, proposal_id);
+        if (should_pass) {
+            proposal.yes_votes = proposal.yes_votes + num_votes;
+        } else {
+            proposal.no_votes = proposal.no_votes + num_votes;
+        };
+
+        Event::emit_event<VoteEvent>(
+            &mut voting_forum.events.vote_events,
+            VoteEvent { proposal_id, num_votes },
+        );
+    }
+
+    /// Resolve the proposal with given id. Can only be done if there are at least as many votes as min required and
+    /// there are more yes votes than no. If either of these conditions is not met, this will revert.
+    ///
+    /// @param voting_forum_address The address of the forum where the proposals are stored.
+    /// @param proposal_id The proposal id.
+    public fun resolve<ProposalType: store>(
+        voting_forum_address: address,
+        proposal_id: u64,
+    ): ProposalType acquires VotingForum {
+        let proposal_state = get_proposal_state<ProposalType>(voting_forum_address, proposal_id);
+        assert!(proposal_state == PROPOSAL_STATE_SUCCEEDED, Errors::invalid_argument(EPROPOSAL_CANNOT_BE_RESOLVED));
+        let voting_forum = borrow_global_mut<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = Table::borrow_mut(&mut voting_forum.proposals, proposal_id);
+
+        let execution_hash = proposal.execution_hash;
+        if (Option::is_some(&execution_hash)) {
+            assert!(
+                TransactionContext::get_script_hash() == *Option::borrow(&execution_hash),
```

Comment:
I don't understand this line. `TransactionContext::get_script_hash()` returns the txn_data_hash of the txn calling the resolve. why do we expect it to be same as execution_hash?

Reply:
It is the script hash. See my comment earlier in Rust.

Reply:
`TransactionContext::get_script_hash()` is the script hash of txn call the resolve.
`execution_hash` is script_hash of TXN making the actually update. `execution_hash` is what is proposed and why it is expected equal with a hash containing call to resolve? I guess I may lack some context here

Reply:
It should be the same thing. When the proposal is created, it should pass the hash of the RESOLUTION script, which is the script that will call resolve and make the actual changes, not the original transaction that creates the proposal.